src/main/scala/viper/gobra/reporting/StatsCollector.scala

```diff
@@ -0,0 +1,376 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.reporting
+
+import org.apache.commons.io.FileUtils
+import org.bitbucket.inkytonik.kiama.relation.NodeNotInTreeException
+import viper.gobra.ast.frontend.{PDomainFunction, PExpression, PFPredicateDecl, PFunctionDecl, PMPredicateDecl, PMPredicateSig, PMethodDecl, PMethodImplementationProof, PMethodSig, PNode, PParameter, PPredConstructor}
+import viper.gobra.ast.internal.BuiltInMember
+import viper.gobra.frontend.{Config, PackageEntry}
+import viper.gobra.frontend.info.{Info, TypeInfo}
+import viper.gobra.util.Violation
+import viper.silver.ast.{FuncApp, Function, Member, Method, MethodCall, Node, Predicate, PredicateAccess}
+import viper.silver.reporter.Time
+
+import scala.collection.concurrent.{Map, TrieMap}
+import java.nio.charset.StandardCharsets.UTF_8
+import java.io.File
+
+object GobraNodeType extends Enumeration {
+  type GobraNodeType = Value
+  val MethodDeclaration, FunctionDeclaration, MethodSignature, FunctionPredicateDeclaration, MethodPredicateDeclaration,
+  MethodImplementationProof, MethodPredicateSignature, DomainFunction, PredicateConstructor = Value
+}
+
+object ViperNodeType extends Enumeration {
+  type ViperNodeType = Value
+  val Function, Predicate, Method = Value
+}
+
+case class StatsCollector(reporter: GobraReporter) extends GobraReporter {
+  import GobraNodeType._
+  import ViperNodeType._
+
+  // Stores type info for a task
+  private[reporting] val typeInfos: Map[String, TypeInfo] = TrieMap()
+  // Maps a gobra member name to a gobra member entry
+  private[reporting] val memberMap: Map[String, GobraMemberEntry] = TrieMap()
+  // Maps a viper member name to a gobra member entry
+  private val viperMemberNameGobraMemberMap: Map[String, GobraMemberEntry] = TrieMap()
+
+  // indentation prefix for generated json
+  private val i = "  "
+  case class GobraMemberEntry(pkgDir: String,
+                              pkg: String,
+                              memberName: String,
+                              args: String,
```

Comment:
what is `args`?

Reply:
It contains a string of the arguments for the gobra member

Reply:
I see. For documentation purposes and reusability, I think it would be better to keep the data as structures as possible, e.g. to use `Vector[PParameter]` instead of a plain string